src/api_strategy.ts

```diff
@@ -0,0 +1,116 @@
+import { CommonDataModel } from "./api_calls"
+import { Cache } from "./cache";
+import { MapCache } from "./map_cache";
+var _ = require('lodash');
+
+/**
+ * An API strategy is an abstract class
+ * meant to provide interface functionality
+ * when getting a response from an API.
+ * Each strategy implements its own version
+ * of getRawResponse, and the APIStrategy
+ * parent abstract class provides context
+ * to the strategy in the form of apiJSON,
+ * and top level mapping methods to convert
+ * the raw response into a useable common
+ * data model.
+ */
+export abstract class APIStrategy {
+    // The JSON for the API
+    protected apiJSON: any;
+
+    // The unique instance ID for the API,
+    // used for caching
+    protected apiID: number;
+
+    // The static counter used for creating a unique API ID
+    private static apiCounter = 0;
+
+    constructor(apiJSON: any) {
+        this.apiJSON = apiJSON;
+        this.apiID = APIStrategy.apiCounter++;
+    }
+
+    /**
+     * Get the common data model response
+     * @param token The parsed text token, for example a URL.
+     */
+    public async getResponse(token: string): Promise<CommonDataModel> {
+        return this.asCommonDataModel(await this.getRawResponse(token));
+    }
+
+    /**
+     * @returns The type of tokens accepted, URL, IP, etc...
+     */
+    public getTokenTypes(): string[] {
+        return this.apiJSON.type;
+    }
+
+    /**
+     * The raw response interface function
+     * @param token The parsed text token, for example a URL.
+     */
+    protected abstract getRawResponse(token: string): Promise<any>;
+
+    /**
+     * Use the api specific mapping from the api json to map from the
+     * api response to the common data model.
+     * @param mapping The mapping from an api response to the common data model, provided in package.json
+     * @param response The raw api response to be mapped
+     */
+    private normalize(mapping: any, response: any): CommonDataModel {
```

Comment:
Is this VirusTotal specific?

Reply:
The implementation itself is not, but the fields Caleb wishes to see are decided based on a VirusTotal response, hence our current CommonDataModel type.

Reply:
There are talks of making the CDM configurable but it currently has a low priority due to the complexity the change might bring.

Reply:
I am more optimistic than zed about making the CDM config-driven, although I do admit that we should have rigorous testing before we do that.